[CI/Build][Docker] Add centralized version manifest for Docker builds

mritunjaysharma394 opened the pull request:
## Purpose

Introduces `docker/versions.json` as a centralized, machine-readable version manifest for all pinned dependencies in Docker builds.

## Problem

Version information is currently scattered across `docker/Dockerfile` and various install scripts, making it hard to:
- Track version changes across releases
- Programmatically query versions for CI/tooling
- Maintain consistency

## Solution

**Dockerfile = Source of Truth** with auto-generated bake-native `versions.json`.

## Integration with Docker Bake (#31477)

This PR complements the `docker buildx bake` work in #31477:

| This PR | Bake PR #31477 |
|---------|----------------|
| `versions.json` = version data | `docker-bake.hcl` = build config |
| What versions to use | How to build |

### Bake-Native JSON Format

```json
{
  "variable": {
    "CUDA_VERSION": { "default": "12.9.1" },
    "PYTHON_VERSION": { "default": "3.12" },
    "torch_cuda_arch_list": { "default": "7.0 7.5 8.0 8.9 9.0 10.0 12.0" },
    "FLASHINFER_VERSION": { "default": "0.5.3" },
    "DEEPGEMM_GIT_REF": { "default": "594953acce41793ae00a1233eb516044d604bcb6" },
    ...
  }
}
```

Bake reads this directly - no preprocessing needed:
```bash
docker buildx bake -f docker/docker-bake.hcl -f docker/versions.json
```

### Auto-Generation

`tools/generate_versions_json.py` parses Dockerfile ARG defaults and generates bake-native versions.json:

```bash
python tools/generate_versions_json.py        # Generate
python tools/generate_versions_json.py --check  # CI validation
```

## Usage

```bash
# Simple docker build (uses Dockerfile defaults)
docker build -f docker/Dockerfile .

# Bake with versions.json
docker buildx bake -f docker/docker-bake.hcl -f docker/versions.json

# Override specific versions
docker buildx bake --set "*.args.CUDA_VERSION=12.8"

# Extract versions programmatically (e.g., for downstream packaging)
jq -r '.variable.CUDA_VERSION.default' docker/versions.json
```

## Benefits

1. **Single Source of Truth**: Dockerfile ARG defaults are authoritative
2. **Bake-Native**: versions.json works directly with `docker buildx bake`
3. **Machine-Readable**: Easy to parse with `jq` for CI and external tooling  
4. **Extension Commits Centralized**: DEEPGEMM_GIT_REF, PPLX_COMMIT_HASH, DEEPEP_COMMIT_HASH moved to Dockerfile
5. **Cleaner Diffs**: Version bumps show as clear changes
6. **Low Barrier**: Contributors can still use `docker build .` without learning bake

## Files Changed

- `docker/versions.json` - Bake-native version manifest (auto-generated)
- `docker/Dockerfile` - Source of truth for version defaults
- `tools/generate_versions_json.py` - Auto-generation script with --check mode

## Test Plan

- [x] JSON is valid and parseable by bake
- [x] `docker buildx bake --print` shows correct variable values
- [x] `python tools/generate_versions_json.py --check` passes
- [x] Pre-commit checks pass

mritunjaysharma394 commented on docker/build.sh:
```diff
@@ -0,0 +1,139 @@
+
+if [ "$DRY_RUN" = true ]; then
+    echo "Dry run - would execute:"
+    echo "${CMD[*]}"
```
Resolved

mritunjaysharma394 committed:
[CI/Build][Docker] Add centralized version manifest for Docker builds

Introduces docker/versions.json as a machine-readable version manifest
for all pinned dependencies used in Docker builds.

This complements the docker buildx bake work in PR #31477:
- versions.json provides the version data (what versions to use)
- docker-bake.hcl provides build configuration (how to build)

Benefits:
- Single source of truth for all pinned versions
- Machine-readable format for CI and external tooling (jq-parseable)
- Cleaner diffs for version bumps
- Easy release comparison: git diff v0.13..v0.14 -- docker/versions.json

Files:
- docker/versions.json: Version manifest with CUDA, Python, FlashInfer,
  bitsandbytes, torch arch list, and extension commit refs
- docker/Dockerfile: Added comments pointing to versions.json

Signed-off-by: Mritunjay Sharma <mritunjay.sharma@chainguard.dev>

mritunjaysharma394 commented:
Thanks for the review! Great questions - I tested these locally and here's what works:

### Q1: How to ingest all build args easily?

**Option A: One-liner to export as env vars**
```bash
eval "$(jq -r '.docker.build_args | to_entries | .[] | select(.key | startswith("_") | not) | "export \(.key)=\"\(.value)\""' docker/versions.json)"
docker buildx bake
```

**Option B: Generate --set flags directly**
```bash
docker buildx bake $(jq -r '.docker.build_args | to_entries | .[] | select(.key | startswith("_") | not) | "--set *.args.\(.key)=\"\(.value)\""' docker/versions.json | tr '\n' ' ')
```

The `select(.key | startswith("_") | not)` filters out the comment fields.

### Q2: How to manage overrides?

Pattern is: **versions.json = defaults, CLI/env = overrides**

```bash
# Use defaults but override specific values
CUDA_VERSION=12.8 docker buildx bake

# Or with --set
docker buildx bake --set "*.args.CUDA_VERSION=12.8"
```

For release pipelines - source defaults from versions.json, then override what you need via env vars or `--set`.

---

Also made a small fix: changed `TORCH_CUDA_ARCH_LIST` to lowercase `torch_cuda_arch_list` in the JSON to match the Dockerfile ARG name (the Dockerfile internally converts it to uppercase `ENV TORCH_CUDA_ARCH_LIST`).

Let me know if you'd like any other patterns documented!

mritunjaysharma394 committed:
[CI/Build][Docker] Add bake-native versions.json with auto-generation

Restructure versions.json to bake-native format per reviewer feedback,
enabling direct use with docker buildx bake without preprocessing.

Changes:
- docker/versions.json: Bake-native format with "variable" structure
- tools/generate_versions_json.py: Auto-generates versions.json from
  Dockerfile ARG defaults, with --check mode for CI validation
- docker/Dockerfile:
  - Updated comments to reflect Dockerfile as source of truth
  - Moved DEEPGEMM_GIT_REF, PPLX_COMMIT_HASH, DEEPEP_COMMIT_HASH
    defaults from install scripts to Dockerfile for centralized
    version management (users can still override via --build-arg)

Usage:
  # Build with bake (no preprocessing needed)
  docker buildx bake -f docker/docker-bake.hcl -f docker/versions.json

  # Query versions programmatically
  jq -r '.variable.CUDA_VERSION.default' docker/versions.json

  # Regenerate after Dockerfile changes
  python tools/generate_versions_json.py

  # CI validation
  python tools/generate_versions_json.py --check

Signed-off-by: Mritunjay Sharma <mritunjay.sharma@chainguard.dev>

mritunjaysharma394 commented:
@Harry-Chen Thanks for the review! Following @amrmahdi's suggestion, we now have a much cleaner approach:

### Q1: How to ingest all build args?

**No preprocessing needed!** The `versions.json` is now bake-native:

```bash
docker buildx bake -f docker/docker-bake.hcl -f docker/versions.json
```

Bake reads `versions.json` directly and applies all variable defaults automatically.

### Q2: How to manage overrides?

```bash
# Override via --set
docker buildx bake --set "*.args.CUDA_VERSION=12.8"

# Or via environment variable
CUDA_VERSION=12.8 docker buildx bake
```

For release pipelines: bake uses `versions.json` defaults, override what you need via `--set` or env vars.

### Extracting versions programmatically

```bash
jq -r '.variable.CUDA_VERSION.default' docker/versions.json
# Output: 12.9.1
```

Much simpler than the previous approach - thanks to Amr's suggestion to use the bake-native JSON format!

mritunjaysharma394 commented on docker/build.sh:
```diff
@@ -0,0 +1,140 @@
+#   -h, --help           Show this help message
+#
+# Examples:
+#   ./docker/build.sh                              # Build default target
```
@amrmahdi Thanks for the excellent suggestion! This is definitely a win win, I've implemented your proposed approach in the latest commit:

Now it will 

1. **Bake-native `versions.json`** - Restructured to the format bake reads directly:
   ```json
   {
     "variable": {
       "CUDA_VERSION": { "default": "12.9.1" },
       "PYTHON_VERSION": { "default": "3.12" },
       ...
     }
   }
   ```
   Now works with: `docker buildx bake -f docker/docker-bake.hcl -f docker/versions.json`

2. **Dockerfile = source of truth** - ARG defaults in Dockerfile are singe source of truth, overides we all can still do

3. **Auto-generation script** - `tools/generate_versions_json.py` parses Dockerfile ARGs and generates bake-native versions.json:
   ```bash
   python tools/generate_versions_json.py        # Generate
   python tools/generate_versions_json.py --check  # CI validation
   ```

4. **Extension commits centralized** - Moved `DEEPGEMM_GIT_REF`, `PPLX_COMMIT_HASH`, `DEEPEP_COMMIT_HASH` defaults from install scripts to Dockerfile for single source of truth (users can still override via `--build-arg` or bake `--set`).

### Usage

```bash
# Simple docker build (uses Dockerfile defaults)
docker build -f docker/Dockerfile .

# Bake with versions.json
docker buildx bake -f docker/docker-bake.hcl -f docker/versions.json

# Override specific versions
docker buildx bake --set "*.args.CUDA_VERSION=12.8"

# Extract versions programmatically (for downstream packaging)
jq -r '.variable.CUDA_VERSION.default' docker/versions.json
```

Let me know if you'd like any adjustments!

amrmahdi commented:
> @Harry-Chen Thanks for the review! Following @amrmahdi's suggestion, we now have a much cleaner approach:
> 
> ### Q1: How to ingest all build args?
> **No preprocessing needed!** The `versions.json` is now bake-native:
> 
> ```shell
> docker buildx bake -f docker/docker-bake.hcl -f docker/versions.json
> ```
> 
> Bake reads `versions.json` directly and applies all variable defaults automatically.
> 
> ### Q2: How to manage overrides?
> ```shell
> # Override via --set
> --"*.args.CUDA_VERSION=12.8"
> 
> # Or via environment variable
> CUDA_VERSION=12.8 docker buildx bake
> ```
> 
> For release pipelines: bake uses `versions.json` defaults, override what you need via `--set` or env vars.
> 
> ### Extracting versions programmatically
> ```shell
> jq -r '.variable.CUDA_VERSION.default' docker/versions.json
> # Output: 12.9.1
> ```
> 
> Much simpler than the previous approach - thanks to Amr's suggestion to use the bake-native JSON format!

Thanks for the update, a few things:

1. To override an ARG, you only need to do `--set <ARG_NAME>`, the other syntax is not correct I believe.
2. Can we please avoid the regex approach? . I'd swap it out for [dockerfile-parse](https://pypi.org/project/dockerfile-parse/) 
3. The `--check` mode exists but it's not wired up anywhere, is your intention to make is a pre commit hook and run it as part of the CI?
4. To view resolved, one can run `docker buildx bake -f docker/versions.json --print`

mritunjaysharma394 committed:
[CI/Build][Docker] Use dockerfile-parse instead of regex

Per reviewer feedback, replace regex-based ARG parsing with
dockerfile-parse library for more robust Dockerfile handling.

Uses parser.structure to extract ARGs from multi-stage Dockerfiles
since the args property doesn't work reliably with complex builds.

Signed-off-by: Mritunjay Sharma <mritunjay.sharma@chainguard.dev>

mritunjaysharma394 commented:
@amrmahdi Thanks for the feedback! Here's what I've done:

1. Override syntax
You're right - I've updated my previous comments. The correct syntax is:
```bash
docker buildx bake --set CUDA_VERSION=12.8
```
Or via environment variable:
```bash
CUDA_VERSION=12.8 docker buildx bake
```

2. dockerfile-parse is now being used instead of regex
Swapped out the regex approach for `dockerfile-parse` in the latest commit. Uses `parser.structure` to reliably extract ARGs from multi-stage Dockerfiles.

3. --check mode
The intention is to run it in CI to catch drift between Dockerfile and versions.json. Options:
- **CI workflow step**: Add to existing Docker build workflow
- **Pre-commit hook**: Add to `.pre-commit-config.yaml`

Which do you prefer? I can wire it up either way. A pre-commit hook would catch issues earlier (before push), while a CI step is less intrusive for contributors.

Note: `dockerfile-parse` would need to be added to dev dependencies if we go the pre-commit route.

mritunjaysharma394 committed:
[CI/Build][Docker] Use uppercase variable names to match docker-bake.hcl

Map lowercase Dockerfile ARG names to uppercase bake variable names
for TORCH_CUDA_ARCH_LIST, MAX_JOBS, and NVCC_THREADS to match the
variable naming convention in docker-bake.hcl.

This allows versions.json to properly override bake defaults when
used with: docker buildx bake -f docker-bake.hcl -f versions.json

Signed-off-by: Mritunjay Sharma <mritunjay.sharma@chainguard.dev>

mritunjaysharma394 committed:
[CI/Build][Docker] Add pre-commit hook to validate versions.json

Add validate-docker-versions hook that runs --check mode when
docker/Dockerfile or docker/versions.json changes. Catches drift
between Dockerfile ARG defaults and versions.json during development.

Signed-off-by: Mritunjay Sharma <mritunjay.sharma@chainguard.dev>

mritunjaysharma394 committed:
[CI/Build][Docker] Track all Dockerfile ARGs with resolved interpolation

- Extract all ARGs with defaults instead of a hardcoded list
- Resolve variable interpolation (e.g., ${CUDA_VERSION} -> 12.9.1)
- Filter out ARGs with unresolved refs (optional build-time args)
- Zero maintenance: auto-catches new ARGs as they're added

Signed-off-by: Mritunjay Sharma <mritunjay.sharma@chainguard.dev>

ywang96 committed:
Merge branch 'main' into dockerfile-refactor

mritunjaysharma394 committed:
Merge branch 'main' into dockerfile-refactor

khluu committed:
Merge branch 'main' into dockerfile-refactor

ywang96 committed:
Merge branch 'main' into dockerfile-refactor
